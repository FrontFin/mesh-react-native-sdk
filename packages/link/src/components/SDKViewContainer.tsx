import type { ReactNode } from 'react';
import { View } from 'react-native';
import { DARK_THEME_COLOR_BOTTOM, LIGHT_THEME_COLOR_BOTTOM } from '../constant';

type Props = {
  children: ReactNode
  isDarkTheme?: boolean
};

export const SDKViewContainer = ({ children, isDarkTheme }: Props) => {
  return (
    <View style={{ flex: 1, backgroundColor: isDarkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM }} testID={'link-connect-view-component'}>
      {children}
    </View>
  );
};
