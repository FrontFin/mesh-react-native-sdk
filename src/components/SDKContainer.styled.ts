/* istanbul ignore file */
import { StyleSheet, useColorScheme } from 'react-native';

export const styles = StyleSheet.create({
  navBarContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 64,
    paddingTop: 12,
    paddingStart: 6,
    paddingEnd: 8,
  },
  navBarImgContainer: {
    width: 40,
    height: 40,
    alignItems: 'center',
    justifyContent: 'center',
  },
  navBarImgButton: {
    width: 24,
    height: 24,
    resizeMode: 'contain',
  },
  navBarLogo: {
    height: 18,
    resizeMode: 'contain',
  },
});

export const isDarkMode = useColorScheme ? useColorScheme() === 'dark' : false;

export const statusBarStyle = {
  backgroundColor: isDarkMode ? '#000000' : '#ffffff',
};
