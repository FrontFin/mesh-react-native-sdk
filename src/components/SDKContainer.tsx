import React from 'react';
import type { ReactNode } from 'react';
import { SafeAreaView, StatusBar, View } from 'react-native';
import { useSDKStyles } from './SDKContainer.styled';
import { DARK_THEME_COLOR_TOP, LIGHT_THEME_COLOR_TOP, DARK_THEME_COLOR_BOTTOM, LIGHT_THEME_COLOR_BOTTOM } from '../constant';

type Props = {
  children: ReactNode
  isDarkTheme: boolean
}

export const SDKContainer = ({ children, isDarkTheme }: Props) => {
  const { statusBarStyle } = useSDKStyles(isDarkTheme);
  const topColor = isDarkTheme ? DARK_THEME_COLOR_TOP : LIGHT_THEME_COLOR_TOP;
  const bottomColor = isDarkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM;

  return (
    <View style={{ flex: 1 }} testID={'link-connect-component'}>
      <SafeAreaView style={{ flex: 0, backgroundColor: topColor }} />
      <SafeAreaView style={{ flex: 1, backgroundColor: bottomColor }}>
        <StatusBar backgroundColor={statusBarStyle.backgroundColor} />
        <View style={{ flex: 1 }}>
          {children}
        </View>
      </SafeAreaView>
      <SafeAreaView style={{ flex: 0, backgroundColor: bottomColor }} />
    </View>
  );
};