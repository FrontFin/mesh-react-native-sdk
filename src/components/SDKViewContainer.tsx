import React from 'react';
import type { ReactNode } from 'react';
import { View } from 'react-native';

type Props = {
  children: ReactNode
}

export const SDKViewContainer = ({ children }: Props) => {
  return (
    <View style={{ flex: 1 }} testID={'link-connect-component'}>
      {children}
    </View>
  );
};
