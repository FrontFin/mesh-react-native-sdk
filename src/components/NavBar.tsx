import { useSDKStyles } from './SDKContainer.styled';
import { Image, TouchableOpacity, View } from 'react-native';

type NavBarProps = {
  goBack: () => void;
  showCloseAlert: () => void;
}

export const NavBar = ({goBack, showCloseAlert}: NavBarProps) => {
  const {navBarStyles: styles} = useSDKStyles();

  return (
    <View testID={'native-navbar'} style={styles.navBarContainer}>
      <TouchableOpacity
        testID={'nav-back-button'}
        onPress={goBack} style={styles.navBarImgContainer}>
        <Image
          source={require('../assets/ic_back.png')}
          style={styles.navBarImgButton}
        />
      </TouchableOpacity>

      <TouchableOpacity
        onPress={showCloseAlert}
        style={styles.navBarImgContainer}
        testID={'close-button'}
      >
        <Image
          source={require('../assets/ic_close.png')}
          style={styles.navBarImgButton}
        />
      </TouchableOpacity>
    </View>
  );
}
