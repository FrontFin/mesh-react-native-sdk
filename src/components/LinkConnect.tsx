import { View } from 'react-native';
import { WebView } from 'react-native-webview';
import React, { useMemo, useRef, useState } from 'react';

import { NavBar } from './NavBar';
import { SDKContainer } from './SDKContainer';
import { SDKViewContainer } from './SDKViewContainer';

import type { LinkConfiguration } from '../';
import { useSDKCallbacks } from '../hooks/useSDKCallbacks';
import { sdkSpecs } from '../utils/sdkConfig';
import { DARK_THEME_COLOR_BOTTOM, LIGHT_THEME_COLOR_BOTTOM } from '../constant';

export const LinkConnect = (props: LinkConfiguration) => {
  const {
    showNativeNavbar,
    showWebView,
    linkUrl,
    darkTheme,
    handleMessage,
    handleNavState,
    showCloseAlert,
  } = useSDKCallbacks(props);
  const webViewRef = useRef<WebView>(null);
  const goBack = () => webViewRef?.current?.goBack();

  const injectedScript = useMemo(() => {
    let sdkTypeScript = `
    window.meshSdkPlatform='${sdkSpecs.platform}';
    window.meshSdkVersion='${sdkSpecs.version}';
  `;

    if (props.settings) {
      sdkTypeScript += `
        window.accessTokens='${JSON.stringify(
          props.settings.accessTokens || {}
        )}';
        window.transferDestinationTokens='${JSON.stringify(
          props.settings.transferDestinationTokens || {}
        )}';
      `;
    }

    return sdkTypeScript;
  }, [props.settings]);

  const SDKWrapperComponent = props.renderViewContainer
    ? SDKViewContainer
    : SDKContainer;

  const LoadingComponentWebview = () => {
    return (
      <View style={{
        position: 'absolute',
        zIndex: 10,
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        backgroundColor: darkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM
      }}/>
    );
  };

  const [initialLoading, setInitialLoading] = useState(true);

  if (darkTheme === undefined) {
    return null
  }
  
  return (
    <SDKWrapperComponent isDarkTheme={darkTheme}>
      {showNativeNavbar && (
        <NavBar goBack={goBack} showCloseAlert={showCloseAlert} isDarkTheme={darkTheme} />
      )}
      {initialLoading && <LoadingComponentWebview />}
      {showWebView && linkUrl && (
        <WebView
          bounces={false}
          style={{ backgroundColor: darkTheme ? DARK_THEME_COLOR_BOTTOM : LIGHT_THEME_COLOR_BOTTOM }}
          testID={'webview'} ref={webViewRef}
          source={{ uri: linkUrl }}
          cacheMode={'LOAD_NO_CACHE'}
          onMessage={handleMessage}
          onLoadEnd={() => {
            setInitialLoading(false);
          }}
          startInLoadingState={true}
          javaScriptEnabled={true}
          injectedJavaScript={injectedScript}
          onNavigationStateChange={handleNavState}
        />
      )}
    </SDKWrapperComponent>
  );
};
