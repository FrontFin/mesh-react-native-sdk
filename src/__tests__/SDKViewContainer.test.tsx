import React from 'react';
import { render } from '@testing-library/react-native';

const mockedUseColorScheme = jest.fn();

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => {
  return {
    default: mockedUseColorScheme,
  };
});

import { SDKViewContainer } from '../components/SDKViewContainer';
import { View } from 'react-native';

describe('SDKViewContainer', () => {
  mockedUseColorScheme.mockReturnValue('light');

  test('renders correctly', () => {
    const component = render(<SDKViewContainer>SDKViewContainer Content</SDKViewContainer>);
    const { getByTestId } = component;

    expect(getByTestId('link-connect-view-component')).toBeTruthy();

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  });



  test('renders child component correctly', () => {
    const component = render(<SDKViewContainer><View testID={"child-element"}>Test</View></SDKViewContainer>);
    const { getByTestId } = component;

    expect(getByTestId('link-connect-view-component')).toBeTruthy();
    expect(getByTestId('child-element')).toBeTruthy();
  });
});
