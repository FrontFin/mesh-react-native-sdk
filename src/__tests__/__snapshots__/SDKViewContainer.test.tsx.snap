// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SDKViewContainer renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FBFBFB",
      "flex": 1,
    }
  }
  testID="link-connect-view-component"
>
  SDKViewContainer Content
</View>
`;
